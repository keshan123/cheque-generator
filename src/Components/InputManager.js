import React from "react";
import FormErrors from "./FormErrors";
import Cheque from "./Cheque";
import moment from 'moment';

class InputManager extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      payee: "",
      amount: "",
      date: "",
      formErrors: { payee: "", amount: "", date: "" },
      payeeValid: false,
      amountValid: false,
      dateValid: false,
      formValid: false,
      error: "",
      isLoaded: false,
      valueInString: "",
      reformattedDate: [],
      buttonText: "Submit",
    };
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleUserInput(e) {
    const name = e.target.name;
    const value = e.target.value;
    this.setState({ [name]: value }, () => {
      this.validateField(name, value);
    });
  }

  async handleSubmit() {
    this.setState({ buttonText: 'Loading'})
    await fetch(
      `https://te5299oebg.execute-api.us-west-2.amazonaws.com/prod/${this.state.amount}`
    )
      .then(response => response.text())
      .then(response =>
        this.setState({ isLoaded: true, valueInString: response })
      );
  }

  validateField = (fieldName, value) => {
    let fieldValidationErrors = this.state.formErrors;
    let payeeValid = this.state.payeeValid;
    let amountValid = this.state.amountValid;
    let dateValid = this.state.dateValid;

    switch (fieldName) {
      case "payee":
        if (value === "") {
          payeeValid = false;
          fieldValidationErrors.payee = payeeValid ? "" : " cannot be empty";
        } else {
          payeeValid = true;
          fieldValidationErrors.payee = "";
        }
        break;
      case "amount":
        if (value < 1) {
          if (value > 0) {
            amountValid = true;
            fieldValidationErrors.amount = "";
          } else {
            amountValid = false;
            fieldValidationErrors.amount = amountValid
              ? ""
              : " invalid - (must be greater than $0 and contain no special characters)";
          }
        } else if (value > 10000000000000) {
          amountValid = false;
          this.setState({ amount: 10000000000000 });
          fieldValidationErrors.amount = amountValid
            ? ""
            : " must be less that $ 10000000000000";
        } else {
          amountValid = true;
          fieldValidationErrors.amount = "";
        }
        break;
      case "date":
        if (!value) {
          dateValid = false;
          fieldValidationErrors.date = dateValid ? "" : " cannot be empty";
        } else {
          if (/[a-zA-Z]/.test(value)) {
            dateValid = false;
            fieldValidationErrors.date = dateValid ? "" : " invalid format";
          } else {
            const momentDate = moment(value, 'DD/MM/YYYY', true)
            let reformatDate = momentDate.format('YYYYMMDD');
            const validatedMoment = momentDate.isValid();
            if (!validatedMoment) {
              dateValid = false;
              fieldValidationErrors.date = dateValid ? "" : " not valid";
            } else {
                const splitDate = reformatDate.split('');
                this.setState({ reformattedDate: splitDate})
                dateValid = true;
                fieldValidationErrors.date = "";
            }
          }
        }
        break;
      default:
        break;
    }

    this.setState(
      {
        formErrors: fieldValidationErrors,
        payeeValid: payeeValid,
        amountValid: amountValid,
        dateValid: dateValid
      },
      this.validateForm
    );
  };

  validateForm = () => {
    this.setState({
      formValid:
        this.state.payeeValid && this.state.amountValid && this.state.dateValid
    });
  };

  render() {
    return (
      <>
        {!this.state.isLoaded ? (
          <div className="details-container">
            <label className="form-input__label">Payee</label>
            <input
              className={`form-input ${this.state.formErrors.payee &&
                "invalid"}`}
              type="text"
              value={this.state.payee}
              onChange={event => this.handleUserInput(event)}
              name="payee"
            ></input>
            <label className="form-input__label">Amount in $</label>
            <input
              className={`form-input ${this.state.formErrors.amount &&
                "invalid"}`}
              type="number"
              value={this.state.amount}
              onChange={event => this.handleUserInput(event)}
              name="amount"
            ></input>
            <label className="form-input__label">Date</label>
            <input
              className={`form-input ${this.state.formErrors.date &&
                "invalid"}`}
              value={this.state.date}
              placeholder="DD/MM/YYYY"
              onChange={event => this.handleUserInput(event)}
              name="date"
            ></input>
            <div>
              <FormErrors formErrors={this.state.formErrors} />
            </div>
            <button
              type="submit"
              disabled={!this.state.formValid}
              onClick={this.handleSubmit}
              className={`form-input__submit-button outline ${
                this.state.formValid ? "white-green" : "white-grey"
              }`}
            >
              {this.state.buttonText}
            </button>
          </div>
        ) : (
          <Cheque
            payee={this.state.payee}
            amount={this.state.amount}
            valueInString={this.state.valueInString}
            date={this.state.reformattedDate}
          />
        )}
      </>
    );
  }
}

export default InputManager;
