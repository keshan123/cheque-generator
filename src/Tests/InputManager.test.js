import React from 'react';

import InputManager from '../Components/InputManager';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer
    .create(<InputManager />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('<InputManager />', () => {
  let inputManager;
  beforeEach(() => {
    inputManager = new InputManager({});
  });

  describe('handleUserInput()', () => {
    it('should call setState once in with correct value', () => {
      inputManager.setState = jest.fn();
      const mockValue = 'mockValue'
      const mockEvent = { target: { value: mockValue, name: 'mockname'}}
      inputManager.handleUserInput(mockEvent);

      expect(inputManager.setState.mock.calls.length).toBe(1);
      expect(inputManager.setState.mock.calls[0][0].mockname).toBe('mockValue');
    });

  });

  describe('validateField()', () => {
    it('should set payeeValid to false when no payee value given', () => {
      inputManager.setState = jest.fn();
      inputManager.validateField('payee', '');

      expect(inputManager.setState.mock.calls.length).toBe(1);
      expect(inputManager.setState.mock.calls[0][0].formErrors.payee).toBe(' cannot be empty');
      expect(inputManager.setState.mock.calls[0][0].payeeValid).toBe(false);
    });

    it('should set payeeValid to true when a payee value given', () => {
      inputManager.setState = jest.fn();
      inputManager.validateField('payee', 'mockValue');

      expect(inputManager.setState.mock.calls.length).toBe(1);
      expect(inputManager.setState.mock.calls[0][0].formErrors.payee).toBe('');
      expect(inputManager.setState.mock.calls[0][0].payeeValid).toBe(true);
    });

    it('should set amountValid to true when a amount value given that is less that 1 but greater than 0', () => {
      inputManager.setState = jest.fn();
      inputManager.validateField('amount', 0.5);

      expect(inputManager.setState.mock.calls.length).toBe(1);
      expect(inputManager.setState.mock.calls[0][0].formErrors.amount).toBe('');
      expect(inputManager.setState.mock.calls[0][0].amountValid).toBe(true);
    });

    it('should set amountValid to false when a amount value given that is less that 1 and less than 0', () => {
      inputManager.setState = jest.fn();
      inputManager.validateField('amount', 0);

      expect(inputManager.setState.mock.calls.length).toBe(1);
      expect(inputManager.setState.mock.calls[0][0].formErrors.amount).toBe(' must be greater than $ 0');
      expect(inputManager.setState.mock.calls[0][0].amountValid).toBe(false);
    });

    it('should set amountValid to false when a amount value given exceeds 10000000000000', () => {
      inputManager.setState = jest.fn();
      inputManager.validateField('amount', 10000000000001);

      expect(inputManager.setState.mock.calls.length).toBe(2);
      expect(inputManager.setState.mock.calls[0][0].amount).toBe(10000000000000);
      expect(inputManager.setState.mock.calls[1][0].formErrors.amount).toBe(' must be less that $ 10000000000000');
      expect(inputManager.setState.mock.calls[1][0].amountValid).toBe(false);
    });

    it('should set amountValid to true when a amount value given is valid', () => {
      inputManager.setState = jest.fn();
      inputManager.validateField('amount', 123);

      expect(inputManager.setState.mock.calls.length).toBe(1);
      expect(inputManager.setState.mock.calls[0][0].formErrors.amount).toBe('');
      expect(inputManager.setState.mock.calls[0][0].amountValid).toBe(true);
    });

    it('should set dateValid to false when a date value given is empty (invalid)', () => {
      inputManager.setState = jest.fn();
      inputManager.validateField('date');

      expect(inputManager.setState.mock.calls.length).toBe(1);
      expect(inputManager.setState.mock.calls[0][0].formErrors.date).toBe(' invalid date');
      expect(inputManager.setState.mock.calls[0][0].dateValid).toBe(false);
    });

    it('should set dateValid to true when a date value given valid', () => {
      inputManager.setState = jest.fn();
      inputManager.validateField('date', '2020-10-10');

      expect(inputManager.setState.mock.calls.length).toBe(1);
      expect(inputManager.setState.mock.calls[0][0].formErrors.date).toBe('');
      expect(inputManager.setState.mock.calls[0][0].dateValid).toBe(true);
    });
  });

  describe('validateForm()', () => {
    it('should call setState once', () => {
      inputManager.state.payeeValid = true;
      inputManager.state.amountValid = true;
      inputManager.state.dateValid = true;
      inputManager.setState = jest.fn();

      inputManager.validateForm();

      expect(inputManager.setState.mock.calls.length).toBe(1);
      expect(inputManager.setState.mock.calls[0][0].formValid).toBe(true);
    });
  });
});
