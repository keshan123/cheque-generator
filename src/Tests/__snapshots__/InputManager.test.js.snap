// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="details-container"
>
  <label
    className="form-input__label"
  >
    Payee
  </label>
  <input
    className="form-input "
    name="payee"
    onChange={[Function]}
    type="text"
    value=""
  />
  <label
    className="form-input__label"
  >
    Amount in $
  </label>
  <input
    className="form-input "
    name="amount"
    onChange={[Function]}
    type="number"
    value=""
  />
  <label
    className="form-input__label"
  >
    Date
  </label>
  <input
    className="form-input "
    name="date"
    onChange={[Function]}
    type="date"
    value=""
  />
  <div>
    <div
      className="form-input__error-message"
    >
      
      
      
    </div>
  </div>
  <button
    className="form-input__submit-button outline white-grey"
    disabled={true}
    onClick={[Function]}
    type="submit"
  >
    Submit
  </button>
</div>
`;
